README
rabbit_fcgi: 

It is a simple, multi-processes, and multi-thread c++11 framework for developing c++ fastcgi applications.

Author: takinwang
Email : wxz@cnic.cn

You can use and modify it as you want. 
If you have any question, please let me known!  

How to use it?

1. add location in nginx.conf:

	location /testcgi { 
	    fastcgi_pass   127.0.0.1:9000;  
	    include        fastcgi_params;
	}

2. change line 30 in src/fcgi.hpp  as the following:

#define nginx_context_path "^/testcgi"
	
3. copy src/endp0_impl.hpp to src/test_impl.hpp

   add #include "test_impl.hpp" in the main.cpp like line 27.

   change the namespace endpoint0 to a new namespace what you wanted in src/test_impl.hpp
   
   change uri endpoint in line 18 as you want: 
   		auto & t0 = fcgi::Endpoints[nginx_context_path "/t0.*"];
   		
   edit lambda function of "on_call" in src/test_impl.hpp as you want.
   		[]( fcgi:: request & req, fcgi::response & resp) {
   		  modify as you want ...
   		}
   		
   see fcgi::request and fcgi::response for details. 
   		
see INSTALL to build this project.

How to start it?

./rabbit_fcgi -T 4 -F 4 -d -p 9000





NEWS
Sample NEWS file for conjson project.


src/fcgi.hpp
/*
 * fcgi.hpp
 *
 *  Created on: Apr 1, 2016
 *      Author: root
 */

#ifndef FCGI_HPP_
#define FCGI_HPP_

#include "Utils.hpp"
#include <fstream>
#include <map>

#include <sys/types.h>
#include <sys/time.h>
#include <sys/stat.h>

#include<sys/wait.h>

#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>

#define FCGI_LISTENSOCK_FILENO 0
#define nginx_context_path "^/covjson"

# include <sys/socket.h>
# include <sys/ioctl.h>
# include <netinet/in.h>
# include <netinet/tcp.h>
# include <sys/un.h>
# include <arpa/inet.h>

# include <netdb.h>

#include <fcgiapp.h>
//#include <fcgi_stdio.h>

namespace fcgi {

//#define  FCGI_printf(...) FCGX_FPrintF(request->out, __VA_ARGS__)

#define  FCGI_ENVP(VAR, KEY) { \
	const char * s = FCGX_GetParam(KEY, cgi_request_.envp);  \
	VAR = (s == NULL) ? std::string(): std::string(s); \
}

class request;
class response;

class request {
	private:
		std::unordered_map<std::string, std::string> querys;
		std::string null_string;
		FCGX_Request &cgi_request_;
	public:
		request(FCGX_Request &cgi_request_)
				: cgi_request_(cgi_request_) {

			FCGI_ENVP(DOCUMENT_ROOT, "DOCUMENT_ROOT");
			FCGI_ENVP(HTTP_COOKIE, "HTTP_COOKIE");
			FCGI_ENVP(HTTP_HOST, "HTTP_HOST");
			FCGI_ENVP(HTTP_REFERER, "HTTP_REFERER");
			FCGI_ENVP(HTTP_USER_AGENT, "HTTP_USER_AGENT");
			FCGI_ENVP(HTTPS, "HTTPS");
			FCGI_ENVP(PATH, "PATH");
			FCGI_ENVP(QUERY_STRING, "QUERY_STRING");
			FCGI_ENVP(REMOTE_ADDR, "REMOTE_ADDR");
			FCGI_ENVP(REMOTE_HOST, "REMOTE_HOST");
			FCGI_ENVP(REMOTE_PORT, "REMOTE_PORT");
			FCGI_ENVP(REMOTE_USER, "REMOTE_USER");
			FCGI_ENVP(REQUEST_METHOD, "REQUEST_METHOD");
			FCGI_ENVP(REQUEST_URI, "REQUEST_URI");
			FCGI_ENVP(SCRIPT_FILENAME, "SCRIPT_FILENAME");
			FCGI_ENVP(SCRIPT_NAME, "SCRIPT_NAME");
			FCGI_ENVP(SERVER_ADMIN, "SERVER_ADMIN");
			FCGI_ENVP(SERVER_NAME, "SERVER_NAME");
			FCGI_ENVP(SERVER_SOFTWARE, "SERVER_SOFTWARE");

			QUERY_STRING = Utils::uri_decode(QUERY_STRING);
			REQUEST_URI = Utils::uri_decode(REQUEST_URI);

			size_t pos = REQUEST_URI.find("?");
			if (pos > 0) REQUEST_URI = REQUEST_URI.substr(0, pos);

			Utils::parse_querys(querys, QUERY_STRING);
		}

		const std::string & get_parameter(const std::string & key) {
			return Utils::get_map_value(querys, key, null_string);
		}

	public:
		std::string DOCUMENT_ROOT;
		std::string HTTP_COOKIE;
		std::string HTTP_HOST;
		std::string HTTP_REFERER;
		std::string HTTP_USER_AGENT;
		std::string HTTPS;
		std::string PATH;
		std::string QUERY_STRING;
		std::string REMOTE_ADDR;
		std::string REMOTE_HOST;
		std::string REMOTE_PORT;
		std::string REMOTE_USER;
		std::string REQUEST_METHOD;
		std::string REQUEST_URI;
		std::string SCRIPT_FILENAME;
		std::string SCRIPT_NAME;
		std::string SERVER_ADMIN;
		std::string SERVER_NAME;
		std::string SERVER_SOFTWARE;

};

class response {
	private:
		std::unordered_map<int, std::string> status_lines;
		FCGX_Request &cgi_request_;
		request & request_;
		bool commited;
	public:
		response(FCGX_Request &cgi_request_, request & request)
				: cgi_request_(cgi_request_), request_(request), commited(false), status(200) {

			status_lines[200] = "OK";
			status_lines[201] = "Created";
			status_lines[202] = "Accepted";
			status_lines[204] = "No Content";
			status_lines[206] = "Partial Content";
			status_lines[301] = "Moved Permanently";
			status_lines[302] = "Moved Temporarily";
			status_lines[303] = "See Other";
			status_lines[304] = "Not Modified";
			status_lines[307] = "Temporary Redirect";
			status_lines[400] = "Bad Request";
			status_lines[401] = "Unauthorized";
			status_lines[402] = "Payment Required";
			status_lines[403] = "Forbidden";
			status_lines[404] = "Not Found";
			status_lines[405] = "Not Allowed";
			status_lines[406] = "Not Acceptable";
			status_lines[408] = "Request Time-out";
			status_lines[409] = "Conflict";
			status_lines[410] = "Gone";
			status_lines[411] = "Length Required";
			status_lines[412] = "Precondition Failed";
			status_lines[413] = "Request Entity Too Large";
			status_lines[414] = "Request-URI Too Large";
			status_lines[415] = "Unsupported Media Type";
			status_lines[416] = "Requested Range Not Satisfiable";
			status_lines[500] = "Internal Server Error";
			status_lines[501] = "Not Implemented";
			status_lines[502] = "Bad Gateway";
			status_lines[503] = "Service Temporarily Unavailable";
			status_lines[504] = "Gateway Time-out";
			status_lines[507] = "Insufficient Storage";

			headers["Content-type"] = "text/html;charset=UTF-8";
		}
		~response() {
			commit();
			if (fstream.is_open()) fstream.close();
		}

		void commit() {
			if (commited == true) return;
			commited = true;

			if (status_lines.find(status) == status_lines.end()) {
				status = 500;
			}

			std::string message = status_lines[status];
			std::string message_str = (boost::format("%s %s") % status % message.c_str()).str();
			headers["Status"] = message_str;

			for (auto iter : headers) {
				FCGX_FPrintF(cgi_request_.out, "%s: %s\r\n", iter.first.c_str(), iter.second.c_str());
			}

			FCGX_FPrintF(cgi_request_.out, "\r\n");
			if (status >= 300) {
				if (status >= 400) FCGX_FPrintF(cgi_request_.out, message_str.c_str());
				return;
			}

			if (fstream.is_open()) {
				char buf[8193];

				while (fstream.good() == true) {
					fstream.read(&buf[0], 8192);
					size_t ncount = fstream.gcount();
					if (ncount == 0) {
						break;
					}
					FCGX_FPrintF(cgi_request_.out, &buf[0], ncount);
				}
				return;
			}

			std::string data = sstream.str();
			if (data.size() > 0) {
				FCGX_FPrintF(cgi_request_.out, data.c_str());
			}
		}

	public:
		std::unordered_map<std::string, std::string> headers;

		size_t status;
		std::string message;

		std::stringstream sstream;
		std::ifstream fstream;
};

class Endpoint {
	public:
		std::function<void(request &, response &)> on_call;
};

static std::map<std::string, Endpoint> Endpoints;

}

#endif /* FCGI_HPP_ */


src/Utils.hpp
/*
 * Utils.h
 *
 *  Created on: Mar 18, 2016
 *      Author: root
 */

#ifndef UTILS_H_
#define UTILS_H_

#include <boost/archive/iterators/base64_from_binary.hpp>
#include <boost/archive/iterators/transform_width.hpp>
#include <boost/random/random_device.hpp>
#include <boost/random/uniform_int_distribution.hpp>

#include <unordered_map>
#include <boost/algorithm/string.hpp>
#include <boost/format.hpp>
#include <sstream>
#include <iostream>

#include <stdlib.h>


static const char uri_safe[] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
		0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1,
		1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1,
		1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
		0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
		0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
		0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
		0, 0, 0, 0, 0 };

class Utils {

	public:

		template<class m, class k, class v>
		static const v & get_map_value(m maps, const k & key, const v& def) {
			auto ai = maps.find(key);
			if (ai == maps.end()) return def;
			return ai->second;
		}

		static void parse_querys(std::unordered_map<std::string, std::string>& querys,
				const std::string & query_string) {
			std::vector<std::string> parts;
			boost::split(parts, query_string, boost::is_any_of("&"));

			querys.clear();
			for (auto t : parts) {
				size_t p1 = t.size();
				if (p1 > 0) {
					size_t p0 = t.find("=");
					if (p0 == std::string::npos) continue;
					std::string key = t.substr(0, p0);
					std::string val = t.substr(p0 + 1, p1 - p0 - 1);
					querys[key] = val;
					//							std::cout << key << "\t" << val << std::endl;
				}
			}
		}

//		static void Base64Encode(const std::string& input, std::stringstream& result);

		static std::string gen_password() {
			std::string chars("abcdefghijklmnopqrstuvwxyz"
					"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
					"1234567890"
					"_");

			boost::random::random_device rng;
			boost::random::uniform_int_distribution<int> index_dist(0, chars.size() - 1);

			char s[9];
			for (int i = 0; i < 8; ++i) {
				s[i] = chars[index_dist(rng)];
			}
			s[8] = '\0';
			return std::string(&s[0]);
		}

		static std::string uri_encode(const std::string &uri) {
			if (uri.length() == 0) return uri;
			std::string ret;
			const unsigned char *ptr = (const unsigned char *) uri.c_str();
			ret.reserve(uri.length());

			for (; *ptr; ++ptr) {
				if (!uri_safe[*ptr]) {
					char buf[5];
					memset(buf, 0, 5);
#ifdef WIN32
					_snprintf_s(buf, 5, "%%%X", (*ptr));
#else
					snprintf(buf, 5, "%%%X", (*ptr));
#endif
					ret.append(buf);
				}
				else if (*ptr == ' ') {
					ret += '+';
				}
				else {
					ret += *ptr;
				}
			}
			return ret;
		}

		static std::string uri_decode(const std::string &uri) {
			//Note from RFC1630:  "Sequences which start with a percent sign
			//but are not followed by two hexadecimal characters (0-9,A-F) are reserved
			//for future extension"
			if (uri.length() == 0) return uri;

			const unsigned char *ptr = (const unsigned char *) uri.c_str();
			std::string ret;
			ret.reserve(uri.length());
			for (; *ptr; ++ptr) {
				if (*ptr == '%') {
					if (*(ptr + 1)) {
						char a = *(ptr + 1);
						char b = *(ptr + 2);
						if (!((a >= 0x30 && a < 0x40) || (a >= 0x41 && a < 0x47))) continue;
						if (!((b >= 0x30 && b < 0x40) || (b >= 0x41 && b < 0x47))) continue;
						char buf[3];
						buf[0] = a;
						buf[1] = b;
						buf[2] = 0;
						ret += (char) strtoul(buf, NULL, 16);
						ptr += 2;
						continue;
					}
				}
				if (*ptr == '+') {
					ret += ' ';
					continue;
				}
				ret += *ptr;
			}
			return ret;
		}

};

#define GET_MAP_STR_DEF( a, b ) Utils::get_map_value( a, std::string( b ), std::string() );

#define GET_MAP_STR( a, b, c ) Utils::get_map_value( a, std::string( b ), c );

#endif /* UTILS_H_ */


src/endp0_impl.hpp
/*
 * covjson_impl.hpp
 *
 *  Created on: Apr 2, 2016
 *      Author: root
 */

#ifndef ENDP0_IMPL_HPP_
#define ENDP0_IMPL_HPP_

#include "fcgi.hpp"
#include <pthread.h>

namespace endpoint0 {

class endp {
	private:
		static endp obj;
	public:
		endp() {
			auto & t0 = fcgi::Endpoints[nginx_context_path "/t0.*"];
			t0.on_call =
					[]( fcgi:: request & req, fcgi::response & resp) {

						std::string a = req.get_parameter("testid");
						if (a == "q") {
							resp.status = 404;
							return;
						}

						resp.sstream << "<html><head><title>FastCGI Hello!</title></head>";
						resp.sstream << "<body>";
						resp.sstream << "<h1>FastCGI Hello!</h1>";
						resp.sstream << req.REQUEST_URI << "<br>";
						resp.sstream << boost::format(
								"pgid:%s, pid:%d, threadid:%d\n<br><i>%s</i>\n") % getpgrp() % getpid() % pthread_self() % a.c_str();
						resp.sstream << "</body><html>";
						resp.commit();
					};
		}
};

endp endp::obj; //auto register fcgi::Endpoints callback

}

#endif /* ENDP0_IMPL_HPP_ */


src/endp1_impl.hpp
/*
 * covjson_impl.hpp
 *
 *  Created on: Apr 2, 2016
 *      Author: root
 */

#ifndef ENDP1_IMPL_HPP_
#define ENDP1_IMPL_HPP_

#include "fcgi.hpp"
#include <pthread.h>

namespace endpoint1 {

class endp {
	private:
		static endp obj;
	public:
		endp() {
			auto & t0 = fcgi::Endpoints[nginx_context_path "/t1.*"];
			t0.on_call =
					[]( fcgi:: request & req, fcgi::response & resp) {

						std::string a = req.get_parameter("testid");
						if (a == "q") {
							resp.status = 404;
							return;
						}

						resp.sstream << "<html><head><title>FastCGI Hello!</title></head>";
						resp.sstream << "<body>";
						resp.sstream << "<h1>FastCGI Hello!</h1>";
						resp.sstream << req.REQUEST_URI << "<br>";
						resp.sstream << boost::format(
								"pgid:%s, pid:%d, threadid:%d\n<br><i>%s</i>\n") % getpgrp() % getpid() % pthread_self() % a.c_str();
						resp.sstream << "</body><html>";
						resp.commit();
					};
		}
};

endp endp::obj; //auto register fcgi::Endpoints callback

}

#endif /* ENDP0_IMPL_HPP_ */


src/main.cpp
/*
 ============================================================================
 Name        : conjson.cpp
 Author      : 
 Version     :
 Copyright   : Your copyright notice
 Description : Hello World in C++,
 ============================================================================
 */

#include <boost/shared_ptr.hpp>
#include <boost/optional.hpp>
#include <boost/algorithm/string.hpp>
#include <boost/lexical_cast.hpp>
#include <iostream>
#include <fstream>
#include <vector>
#include <boost/regex.hpp>

#include <pthread.h>

#include "Utils.hpp"

#include "fcgiapp.h"
#include "fcgi.hpp"

/************************/
#include "endp0_impl.hpp"
#include "endp1_impl.hpp"
/************************/

const char * pid_file = "/var/run/covjson.pid";

void write_pid() {
	std::ofstream outfile;
	outfile.open(pid_file);

	if (outfile.is_open() == false) {
		std::cerr << "Write pid_file failed:" << pid_file << std::endl;
		return;
	}
	std::string t = boost::lexical_cast<std::string, int>(getpid());
	outfile.write(t.c_str(), t.size());
	outfile.close();
}

void unlink_pid() {
	struct stat st;
	if (stat(pid_file, &st) == 0) unlink(pid_file);
}

void _signal_term(int code) {
	kill(0, SIGTERM); // kill all forks
	unlink_pid();
	exit(0);
}

int stop_app() {
	std::string line;

	std::ifstream infile;
	infile.open(pid_file);

	if (infile.is_open() == false) return -1;
	if (infile.eof() == false) getline(infile, line);
	infile.close();

	struct stat st;
	if (stat(pid_file, &st) == 0) unlink(pid_file);

	boost::trim<std::string>(line);
	if (line.length() == 0) return -1;

	int fork_pid = 0;
	try {
		fork_pid = boost::lexical_cast<int, std::string>(line);
	}
	catch (std::exception &e) {
	}

	if (fork_pid > 0) {
		std::cout << "Sending signal to pid: " << fork_pid << std::endl;

		kill(fork_pid, SIGTERM);
		kill(fork_pid, SIGINT);

		waitpid(fork_pid, NULL, 0);

		std::cout << "Terminated, pid: " << fork_pid << std::endl;
		return 0;
	}

	return -1;
}

static void * run_thread_fcgi(void *a) {
	std::vector<std::pair<boost::regex, fcgi::Endpoint*> > opt_endpoint;

	for (auto& endp : fcgi::Endpoints) {
		opt_endpoint.emplace_back(boost::regex(endp.first), &endp.second);
	}

	FCGX_Request cgi_request_;
	FCGX_InitRequest(&cgi_request_, FCGI_LISTENSOCK_FILENO, 0);

	while (FCGX_Accept_r(&cgi_request_) >= 0) {
		{
			fcgi::request req(cgi_request_);
			fcgi::response resp(cgi_request_, req);

			for (auto& endp : opt_endpoint) {
				boost::smatch path_match;

				if (boost::regex_match(req.REQUEST_URI, path_match, endp.first)) {
					if (endp.second->on_call) endp.second->on_call(req, resp);
					break;
				}
			}

			resp.status = 404;
		}

		FCGX_Finish_r(&cgi_request_);
	}
	return NULL;
}

int run_fcgi(int thrd_count) {

	FCGX_Init();

	pthread_t thread_id[thrd_count];

	for (int i = 0; i < thrd_count; i++)
		pthread_create(&thread_id[i], NULL, run_thread_fcgi, NULL);

	for (int i = 0; i < thrd_count; i++)
		pthread_join(thread_id[i], NULL);

	return 0;
}

int fcgi_spawn_connection(int fork_count, int thrd_count) {
	int rc = 0;
	struct timeval tv = { 0, 100 * 1000 };
	pid_t child;
	int children = 0;

	for (;;) {
		switch (child = fork()) {
		case 0: {
			/* in child */

//			setsid();
//			chdir("/");

			int max_fd = open("/dev/null", O_RDWR);
			if (-1 != max_fd) {
				if (max_fd != STDOUT_FILENO) dup2(max_fd, STDOUT_FILENO);
				if (max_fd != STDERR_FILENO) dup2(max_fd, STDERR_FILENO);
				if (max_fd != STDOUT_FILENO && max_fd != STDERR_FILENO) close(max_fd);
			}
			else {
				std::cerr << "Couldn't open and redirect stdout/stderr to '/dev/null': " << strerror(errno)
						<< std::endl;
			}

			/* we don't need the client socket */
			for (int i = 3; i < max_fd; i++) {
				if (i != FCGI_LISTENSOCK_FILENO) close(i);
			}

			run_fcgi(thrd_count);

			exit(0);
		}
		case -1: {
			/* error */
			std::cerr << "Fork failed: " << strerror(errno) << std::endl;
			exit(0);
		}
		default: {
			/* in parent */

			select(0, NULL, NULL, NULL, &tv);

			int status;
			children++;
			while (children >= fork_count) {
				switch (child = waitpid(WAIT_ANY, &status, 0)) {
				case -1:
					std::cerr << "Waitpid failed: " << strerror(errno) << std::endl;
					exit(0);
				default:
//					fprintf(stdout, "Child spawned waitpid PID: %d\n", child);

					if (WIFEXITED(status)) {
						std::cerr << "Child exited with: " << WEXITSTATUS(status) << std::endl;
						children--;
					}
					else if (WIFSIGNALED(status)) {
						std::cerr << "Child signaled: " << WTERMSIG(status) << std::endl;
						int status0;
						pid_t child0;
						switch (child0 = waitpid(child, &status0, WNOHANG)) {
						case -1: {
							std::cerr << "Waitpid [" << child << "] failed: " << strerror(errno) << std::endl;
							children--;
							break;
						}
						default:
							if (WIFEXITED(status0)) {
								std::cerr << "Child exited with: " << WEXITSTATUS(status0) << std::endl;
								children--;
							}
						}
					}
					else {
						std::cerr << "Child died somehow, exit status: " << status << std::endl;
						children--;
					}
				}

			}
		}
		}
	}

	return rc;
}

int main(int argc, char **argv) {

	if (argc == 2) {
		if (boost::equal<std::string, std::string>(argv[1], "stop")) return stop_app();
	}

	if (stop_app() == 0) {
		sleep(1);
	}

	unsigned short port = 9000;

	int fork_count = 0;
	int thrd_count = 1;
	int daemonized = 0;

	for (int i = 1; i < argc; i++) {
		std::string a_arg = argv[i];
		if (boost::equal<std::string, std::string>(a_arg, "-d")) {
			daemonized = 1;
		}
		else if (boost::equal<std::string, std::string>(a_arg, "-p") && (i < argc - 1)) {
			port = boost::lexical_cast<int, std::string>(argv[++i]);
		}
		else if (boost::equal<std::string, std::string>(a_arg, "-F") && (i < argc - 1)) {
			fork_count = boost::lexical_cast<int, std::string>(argv[++i]);
		}
		else if (boost::equal<std::string, std::string>(a_arg, "-T") && (i < argc - 1)) {
			thrd_count = boost::lexical_cast<int, std::string>(argv[++i]);
		}
		else if (boost::equal<std::string, std::string>(a_arg, "-h")) {
			std::cout << "Example: " << argv[0] << " <-p port> <-F forks> <-T threads> <-d>" << std::endl;
			return -1;
		}
	}

	std::cout << "Start with " << fork_count << " forks and " << thrd_count << " threads each fork, listen on port: "
			<< port << std::endl;

	signal(SIGTERM, _signal_term);
	signal(SIGINT, _signal_term);

	if (daemonized > 0) {
		daemon(0, 0);
	}

	std::string socket_port = (boost::format(":%d") % port).str();
	int fcgi_fd = FCGX_OpenSocket(socket_port.c_str(), 1024);
	if (-1 == fcgi_fd) {
		return -1;
	}

	write_pid();

	if (fcgi_fd != FCGI_LISTENSOCK_FILENO) {
		close(FCGI_LISTENSOCK_FILENO);
		dup2(fcgi_fd, FCGI_LISTENSOCK_FILENO);
		close(fcgi_fd);
	}

	int val = 1;
	if (setsockopt(FCGI_LISTENSOCK_FILENO, SOL_SOCKET, SO_REUSEADDR, &val, sizeof(val)) < 0) {
		std::cerr << "Couldn't set SO_REUSEADDR: " << strerror(errno) << std::endl;
		return -1;
	}

	if (fork_count <= 0) return run_fcgi(thrd_count);

	return fcgi_spawn_connection(fork_count, thrd_count);
}
